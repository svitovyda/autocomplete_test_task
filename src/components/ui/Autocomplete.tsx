import { useDataInputManager } from '../../hooks/dataInputQueryManager';
import { useClickOutside } from '../../hooks/useClickOutside';
import { SelectContainer } from '../../styles/ui/Autocomplete';
import * as React from 'react';

export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface AutocompleteProps<T> {
  onQueryChanged?: (query: string) => void;
  onItemSelected?: (item: T | undefined) => void;
  minAcceptableLength?: number;
  debounceInterval?: number;
  placeholder?: string;
  maxOptionsToShow?: number;
  initialQuery?: string;
  autoCalculate?: boolean;
  data: T[];
  dataLabel?: (item: T) => string;
  dataId?: (item: T) => ItemID;
  caseSensitive?: boolean;
  showInitOnEmptyInput?: boolean;
}
/*
 * This component renders input with autocompletion
 * It can be in two modes:
 * - autoCalculate ("static")
 *   it gets all the data once and then as user interacts, searches through this given data
 * - "dynamic"
 *   it assumes that all the calculations with data search by query happen outside of the component,
 *   and it just renders data
 */
export const Autocomplete = <T,>({
  onQueryChanged,
  onItemSelected,
  minAcceptableLength = 3,
  debounceInterval = 3,
  placeholder = '',
  maxOptionsToShow = 5,
  initialQuery = '',
  autoCalculate,
  data,
  dataLabel,
  dataId,
  caseSensitive = false,
  showInitOnEmptyInput = true,
}: AutocompleteProps<T>) => {
  const [dataInputState, setInput, setSelectedItemId] = useDataInputManager(
    data,
    debounceInterval,
    minAcceptableLength,
    autoCalculate,
    dataLabel,
    dataId,
    caseSensitive,
    showInitOnEmptyInput,
    maxOptionsToShow
  );

  const [showDropDown, setShowDropDown] = React.useState(false);

  const inputRef = React.useRef<HTMLInputElement>(null);
  const selectRef = React.useRef<HTMLSelectElement>(null);

  React.useEffect(() => {
    onItemSelected?.(dataInputState.selectedItem?.itemData);
    if (dataInputState.selectedItem) {
      setInput(dataInputState.selectedItem?.itemToShow.label);
    }
  }, [dataInputState.selectedItem, onItemSelected, setInput]);

  React.useEffect(() => {
    onQueryChanged?.(dataInputState.query);
  }, [dataInputState.query, onQueryChanged]);

  React.useEffect(() => {
    setInput(initialQuery);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const onFocusIn = React.useCallback(() => {
    // show drop-out, data to show have to be set in `handleChange`
    setShowDropDown(true);
    if (inputRef.current && dataInputState.input) {
      inputRef.current.setSelectionRange(0, dataInputState.input.length);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectRef?.current, inputRef?.current, dataInputState.input, dataInputState.itemsToShow]);

  const handleChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = e.target;
      setInput(value);
    },
    [setInput]
  );

  const onClickOutside = React.useCallback(() => {
    setShowDropDown(false);
  }, []);

  useClickOutside([selectRef, inputRef], onClickOutside);

  const onSelectItem = React.useCallback(
    (id: ItemID) => {
      setSelectedItemId(id);
      setShowDropDown(false);
    },
    [setSelectedItemId]
  );

  return (
    <>
      <input
        ref={inputRef}
        type="text"
        value={dataInputState.input}
        onChange={handleChange}
        placeholder={placeholder}
        onFocus={onFocusIn}
      />
      <SelectContainer>
        {showDropDown && dataInputState.itemsToShow.length > 0 && (
          <select
            ref={selectRef}
            onChange={(e: React.ChangeEvent<HTMLSelectElement>) => onSelectItem(e.target.value)}
            size={Math.max(dataInputState.itemsToShow.length, 2)} // with only one item the select height gets too small and behaves weird
          >
            {dataInputState.itemsToShow.map((item) => (
              <option key={item.id} value={item.id}>
                {item.label}
              </option>
            ))}
          </select>
        )}
      </SelectContainer>
    </>
  );
};
