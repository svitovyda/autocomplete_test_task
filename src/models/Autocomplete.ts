export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface DictionaryItem<T> {
  itemToShow: ShowItem;
  itemData: T;
}
