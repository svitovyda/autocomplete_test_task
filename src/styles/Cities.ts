import styled from 'styled-components';

export const AutoselectsContainer = styled.div`
  width: 100%;
  max-width: 600px;
  padding: 0;
  margin: 0;
  display: flex;
  gap: 20px;
`;

export const AutoselectContainer = styled.div`
  box-sizing: border-box;
  flex: 1;
  width: 100%;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
`;

export const Label = styled.h4`
  font-size: 17px;
  font-weight: 400;
  padding: 4px 2px;
  margin: 0;
`;
