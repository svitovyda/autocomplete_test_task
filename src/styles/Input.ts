import styled from 'styled-components';

export const InputContainer = styled.div`
  margin: 8px 0;
  display: block;
  width: 100%;
`;

export const InputStyled = styled.input`
  display: inline-block;
  padding-block: 1px;
  padding-inline: 2px;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 8px;
  outline-width: 0;
  color: inherit;
`;
