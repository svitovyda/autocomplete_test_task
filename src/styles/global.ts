import { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
  html {
    font-family:sans-serif;
    -ms-text-size-adjust:100%;
    -webkit-text-size-adjust:100%;
  }
  body {
    font: 16px "Helvetica Neue",Helvetica,Arial,sans-serif;
    line-height: 1.5;
    padding: 0;
    margin: 0;
    color: #543729
  }
  h1 {
    font-size: 18px;
    font-weight: bold;
    text-align: center;
    padding: 18px 0 4px 0;
    margin: 0;
  }
  input {
    display: inline-block;
    padding: 4px 2px;
    cursor: text;
    border: 0;
    border-radius: 1px;
    box-shadow: 0 0 1px #aaa;
    font-size: 18px;
    font-weight: 400;
    width: 100%;
    margin: 0;
    padding: 8px;
    outline-width: 0;
    color: inherit;
  }
  select {
    box-sizing: border-box;
    appearance: none;
    border-width: 1px;
    border-color: #aaa;
    border-radius: 3px;
    box-shadow: 3px 3px 5px 0 #999;
    position: relative;
    top: 100%;
    left: 0;
    width: 50%;
    min-width: fit-content;
    z-index: 100000;
    background-color: white;
    list-style: 'none';
    padding: 8px;
    margin: 0;
    scrollbar-width: none;
  }
  option {
    box-sizing: border-box;
    margin: 0;
    padding: 4px 2px;
    font-size: 18px;
    font-weight: 400;
    width: auto;
    color: #543729;
    border: '1px solid #ccc';
    border-radius: 1px;
  }
`;
