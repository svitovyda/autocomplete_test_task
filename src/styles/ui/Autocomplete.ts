import styled from 'styled-components';

export const SelectContainer = styled.div`
  width: 100%;
  margin: 0;
  padding: 0;
  display: block;
  height: 0;
`;
