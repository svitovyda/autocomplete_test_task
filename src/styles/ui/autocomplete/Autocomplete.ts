import styled from 'styled-components';

export const AutocompleteInput = styled.input`
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
  ::placeholder {
    color: #aaa;
    font-size: 16px;
  }
`;
