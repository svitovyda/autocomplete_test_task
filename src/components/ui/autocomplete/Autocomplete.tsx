import { type DataArrayType, useDataShow } from '../../../hooks/autocomplete/useDataShow';
import { useInputQuery } from '../../../hooks/autocomplete/useInputQuery';
import { useClickOutside } from '../../../hooks/useClickOutside';
import type { DictionaryItem } from '../../../models/Autocomplete';
import { AutocompleteInput } from '../../../styles/ui/autocomplete/Autocomplete';
import { defaultLabelId } from '../../../utils/autocomplete';
import type { QueryType } from '../../../utils/textSearch';
import { DropDown } from './DropDown';
import _ from 'lodash';
import * as React from 'react';

export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface AutocompleteProps<T> {
  onQueryChanged?: (query: QueryType) => void;
  onItemSelected?: (item: T) => void;
  minAcceptableLength?: number;
  debounceInterval?: number;
  placeholder?: string;
  maxOptionsToShow?: number;
  autoCalculate?: boolean;
  data: DataArrayType<T>;
  dataLabel?: (item: T) => string;
  dataId?: (item: T) => ItemID;
  caseSensitive?: boolean;
  showDataOnEmptyInput?: boolean;
  initialInput?: string;
  loading?: boolean;
  LoadingAnimation?: React.ComponentType<React.HTMLAttributes<HTMLElement>>;
  keepDropDownOnSelect?: boolean;
}
/*
 * This component renders input with autocompletion
 * It can be in two modes:
 * - autoCalculate ("static")
 *   it gets all the data once and then as user interacts, searches through this given data
 * - "dynamic"
 *   it assumes that all the calculations with data search by query happen outside of the component,
 *   and it just renders data
 */
export const Autocomplete = React.memo(
  <T,>({
    onQueryChanged,
    onItemSelected,
    minAcceptableLength = 3,
    debounceInterval = 600,
    placeholder = '',
    maxOptionsToShow = 5,
    autoCalculate,
    data = [],
    dataLabel = defaultLabelId,
    dataId = defaultLabelId,
    caseSensitive = false,
    showDataOnEmptyInput = true,
    initialInput,
    loading,
    LoadingAnimation,
    keepDropDownOnSelect,
  }: AutocompleteProps<T>) => {
    /**
     * If Autocomplete is used to display result of the search that were recalculated outside,
     * `autoCalculate` should be set to false. Then we just always need to convert each item to
     * pair `{id, label}`, and take first `maxOptionsToShow`.
     * So `showInit` should always be `true` when `autoCalculate` is false
     *
     * And only once user clicks on the input or focuses on it, this precalculated pairs should be
     * shown
     *
     * If Autocomplete is using static data and should filter it by itself as user types,
     * `showDataOnEmptyInput` lets us to choose, what to show when user didn't enter the input:
     * no options or just first `maxOptionsToShow` options
     **/

    // const [inputQueryState, setInput, setQuery] = useInputQuery(
    const [inputQueryState, setInput] = useInputQuery(debounceInterval, minAcceptableLength);
    const prevQuery = React.useRef<QueryType>(undefined);

    const [showDropDown, setShowDropDown] = React.useState(false);

    const itemsToShow = useDataShow<T>(
      data,
      dataLabel,
      dataId,
      maxOptionsToShow,
      autoCalculate,
      caseSensitive,
      inputQueryState.query
    );

    const inputRef = React.useRef<HTMLInputElement>(null);
    const dropDownRef = React.useRef<HTMLDivElement>(null);

    React.useEffect(() => {
      if (initialInput !== undefined) {
        setInput(initialInput || '');
      }
    }, [initialInput, setInput]);

    // listens to the query debounced, to make HTTP call for example
    // it should be called no matter if query changed due to user typing or user selected item,
    // we need to make sure external search result has all the correct data from server
    React.useEffect(() => {
      if (inputQueryState.query !== prevQuery.current) {
        prevQuery.current = inputQueryState.query;
        onQueryChanged?.(inputQueryState.query);
      }
    }, [inputQueryState.query, onQueryChanged]);

    // all the calculations of dataToShow should be made by this moment,
    // only show drop-down and select all tezt in input
    const onFocusIn = React.useCallback(() => {
      if (showDataOnEmptyInput || inputQueryState.input) {
        setShowDropDown(true);
        if (inputRef.current && inputQueryState.input) {
          inputRef.current.setSelectionRange(0, inputQueryState.input.length);
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dropDownRef?.current, inputRef?.current, inputQueryState.input, itemsToShow]);

    // listens to user typing
    const handleInputChange = React.useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const { value } = e.target;
        setInput(value);
      },
      [setInput]
    );

    // if user clicks outside of this component - close drop-down
    const onClickOutside = React.useCallback(() => {
      setShowDropDown(false);
    }, []);

    useClickOutside([dropDownRef, inputRef], onClickOutside);

    // listens on user selecting the option from drop-down
    const onSelectItem = React.useCallback(
      (item: DictionaryItem<T>) => {
        onItemSelected?.(item.itemData);
        keepDropDownOnSelect && setShowDropDown(false);
        setInput(item.itemToShow.label);
      },
      [onItemSelected, setInput, keepDropDownOnSelect]
    );

    return (
      <>
        <AutocompleteInput
          ref={inputRef}
          type="text"
          value={inputQueryState.input}
          onChange={handleInputChange}
          placeholder={placeholder}
          onFocus={onFocusIn}
        />

        {showDropDown && itemsToShow.length > 0 && (
          <DropDown
            ref={dropDownRef}
            onSelectItem={onSelectItem}
            items={itemsToShow}
            query={inputQueryState.query}
            caseSensitive={caseSensitive}
            loading={loading}
            LoadingAnimation={LoadingAnimation}
          />
        )}
      </>
    );
  },
  (prevProps: AutocompleteProps<any>, nextProps: AutocompleteProps<any>) => {
    const {
      dataLabel: prevDataLabel,
      dataId: prevDataId,
      onItemSelected: onPrevItemSelected,
      onQueryChanged: onPrevQueryChanged,
      ...prevRest
    } = prevProps;
    const {
      dataLabel: nextDataLabel,
      dataId: nextDataId,
      onItemSelected: onNextItemSelected,
      onQueryChanged: onNextQueryChanged,
      ...nextRest
    } = nextProps;
    return (
      _.isEqual(prevRest, nextRest) &&
      prevDataId === nextDataId &&
      prevDataLabel === nextDataLabel &&
      onPrevItemSelected === onNextItemSelected &&
      onPrevQueryChanged === onNextQueryChanged
    );
  }
);

Autocomplete.displayName = 'Autocomplete';
