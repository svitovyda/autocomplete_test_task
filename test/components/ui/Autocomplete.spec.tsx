import { Autocomplete } from '../../../src/components/ui/Autocomplete';
import { InputStyled } from '../../../src/styles/Input';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';

describe('Autocomplete', () => {
  it('renders without error', () => {
    const component = render(<Autocomplete />);

    expect(component).toMatchSnapshot();
  });

  it('renders without error styled', () => {
    const component = render(<Autocomplete InputComponent={InputStyled} />);

    expect(component).toMatchSnapshot();
  });
});
