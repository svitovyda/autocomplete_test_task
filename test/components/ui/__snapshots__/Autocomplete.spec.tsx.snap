// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  width: 100%;
  margin: 0;
  padding: 0;
  display: block;
  height: 0;
}

<body>
    <div>
      <input
        placeholder=""
        type="text"
        value=""
      />
      <div
        class="c0"
      />
    </div>
  </body>,
  "container": .c0 {
  width: 100%;
  margin: 0;
  padding: 0;
  display: block;
  height: 0;
}

<div>
    <input
      placeholder=""
      type="text"
      value=""
    />
    <div
      class="c0"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
